controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -319,13 +318,43 @@ private ApplicationController applications() {
         return Collections.unmodifiableList(jobs);
     }
 
-    /** Returns whether given job can be triggered at this point in time */
-    private boolean canTrigger(JobType job, Application application, List<JobType> stepJobs, Optional<Instant> completedAt) {
-        if (!completedAt.isPresent()) return false;
+    /** Returns whether given job should be triggered */
+    private boolean canTrigger(JobType job, Application application, List<JobType> stepJobs) {
         if (jobStateOf(application, job) != idle) return false;
-        if (!stepJobs.containsAll(runningProductionJobs(application))) return false;
+        if (stepJobs != null && !stepJobs.containsAll(runningProductionJobs(application))) return false;
 
-        return true;
+        return triggerAt(clock.instant(), job, application);
+    }
+
+    /** Returns whether given job should be triggered */
+    private boolean canTrigger(JobType job, Application application) {
+        return canTrigger(job, application, null);
+    }
+
+    /** Returns whether job can trigger at given instant */
+    private boolean triggerAt(Instant instant, JobType job, Application application) {
+        Optional<JobStatus> jobStatus = application.deploymentJobs().statusOf(job);
+        if (!jobStatus.isPresent()) return true;
+        if (jobStatus.get().isSuccess()) return true; // Success
+        if (!jobStatus.get().lastCompleted().isPresent()) return true; // Never completed
+        if (!jobStatus.get().firstFailing().isPresent()) return true; // Should not happen as firstFailing should be set for an unsuccessful job
```

Comment:
If it shouldn't happen, perhaps `throw new AssertionError()`?

Reply:
It's mostly there to avoid IDE warnings. I assume `firstFailing` being cleared on success is guaranteed elsewhere.